### PR TITLE
Don't add CSRF cookie if the HTTP headers have been written

### LINKED ISSUE #34375: CSRF: head is already written
### Describe the bug

Sometimes I'm getting this exception:

```
2023-06-28 12:26:51,627 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to / failed, error id: f847bb1a-b636-4561-b59b-cc522052a2c0-3: java.lang.IllegalStateException: Response head already sent
	at io.vertx.core.http.impl.Http1xServerResponse.checkHeadWritten(Http1xServerResponse.java:709)
	at io.vertx.core.http.impl.Http1xServerResponse.addCookie(Http1xServerResponse.java:847)
	at io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter.createCookie(CsrfRequestResponseReactiveFilter.java:233)
	at io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter.filter(CsrfRequestResponseReactiveFilter.java:198)
	at io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter$GeneratedServerResponseFilter$filter.filter(Unknown Source)
	at io.quarkus.csrf.reactive.runtime.CsrfRequestResponseReactiveFilter$GeneratedServerResponseFilter$filter_ClientProxy.filter(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.ResourceResponseFilterHandler.handle(ResourceResponseFilterHandler.java:25)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:150)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

I guess this happens if it's too late to write, which is probably fine, and besides there's no point in throwing, the response is already out, we should first check if the response is written and avoid setting cookies.

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by geoand**
Fixes: #34375

### CODE REVIEW

File: extensions/csrf-reactive/runtime/src/main/java/io/quarkus/csrf/reactive/runtime/CsrfRequestResponseReactiveFilter.java [hunk 1]
```diff
@@ -230,7 +230,9 @@ private void createCookie(String csrfToken, RoutingContext routing, CsrfReactive
         if (config.cookieDomain.isPresent()) {
             cookie.setDomain(config.cookieDomain.get());
         }
-        routing.response().addCookie(cookie);
+        if (!routing.response().headWritten()) {
+            routing.response().addCookie(cookie);
```
Comment (gsmet) [inline extensions/csrf-reactive/runtime/src/main/java/io/quarkus/csrf/reactive/runtime/CsrfRequestResponseReactiveFilter.java:234]: I'm not sure but shouldn't this cookie always be written in all requests?

@FroMage you confirm that when you have these errors, you don't care about the CSRF cookie?

**Comment by sberyozkin** [comment]
Thanks @geoand, will that masking some problem though ? I.e, the response filters should be allowed to complete their work - otherwise - if this filter does not create a cookie while the HTML form will have the token embedded, the next POST form request will fail.
@FroMage Is it possible to create a simple reproducer ?

**Comment by FroMage** [comment]
OK, so this happens, in Renarde, when I'm testing authentication failures, by passing an expired or invalid JWT token.

The ONLY way (at the time) that I could register a handler for these authentication failures (that are pretty common in real life) was to introduce a Vert.x route filter to override `QuarkusHttpUser.AUTH_FAILURE_HANDLER`:

```
    @RouteFilter(100)
    void myFilter(RoutingContext rc) {
        rc.put(QuarkusHttpUser.AUTH_FAILURE_HANDLER, new BiConsumer<RoutingContext, Throwable>() {
            @Override
            public void accept(RoutingContext routingContext, Throwable throwable) {
                while (throwable.getCause() != null)
                    throwable = throwable.getCause();
                if (throwable instanceof InvalidJwtException) {
                    InvalidJwtException x = (InvalidJwtException) throwable;
                    if (x.hasErrorCode(ErrorCodes.EXPIRED)) {
                        redirectToRoot(routingContext, "Login expired, you've been logged out");
                        return;
                    }
                    redirectToRoot(routingContext, "Invalid session (bad JWT), you've been logged out");
                }
                // This happens when the private/public keys change, like in DEV mode
                if (throwable instanceof AEADBadTagException
                        || throwable instanceof BadPaddingException) {
                    redirectToRoot(routingContext, "Invalid session (bad signature), you've been logged out");
                    return;
                }
                // let upstream handle the rest
            }

            private void redirectToRoot(RoutingContext routingContext, String message) {
                // FIXME: constant
                routingContext.removeCookie("QuarkusUser").setPath("/");
                Map<String, Object> map = new HashMap<>();
                // FIXME: format?
                map.put("message", message);
                Flash.setFlashCookie(routingContext.request(), routingContext.response(), map);
                // FIXME: URI, perhaps redirect to login page?
                // Note that this calls end()
                routingContext.response().setStatusCode(303);
                routingContext.redirect("/");
            }
        });
        rc.next();
    }
```

So, in this case, the security failure handler has already handled the response and sent the redirect. I'm pretty sure we don't care if the CSRF cookie is added in this case.

Now, if things have changed and I can actually handle these exceptions in regular `ExceptionMapper` that'd be great, and I can test how it behaves, so that's also an alternative.

But I wonder why the stock `QuarkusHttpUser.AUTH_FAILURE_HANDLER` wouldn't trigger the exact same issue when returning HTTP errors?

**Comment by michalvavrik** [comment]
This really depends on exception, we documented that you can use ex. mappers for auth exceptions https://quarkus.io/guides/security-proactive-authentication-concept#how-to-customize-authentication-exception-responses but here you catch `InvalidJwtException`. That one should be failed by default auth failure handler (which you override right now) and it would go to Vert.X failure handlers, so you should be able to define Vert.X failure handler before it gets to fallback Quarkus failure handler.

TBH I'd need reproducer to be sure as **I don't know CSRF well enough**, but if RESTEasy Reactive already started processing request, I am surprised that default auth failure handler is applied as what we do is to remove it and leave default RR exception mappers to handle it (as we do for auth exceptions to avoid these exact situations). What this PR does will avoid ugly logging, but it should not be necessary. And that way, custom exception mappers would work as well.

**Comment by michalvavrik** [comment]
`io.quarkus.oidc.runtime.OidcUtils#validateAndCreateIdentity` the exception should be wrapped with auth exception, so +1 for trying ex. mapper

**Comment by geoand** [comment]
Should I close this or not?

I can't say I am interested in all the details and would be glad to leave them to you folks :)

**Comment by michalvavrik** [comment]
> Should I close this or not?
> 
> I can't say I am interested in all the details and would be glad to leave them to you folks :)

:-D It might come down to your solution, but I would expect things to work differently and checking if you can write to ended event is like fighting hydra -> you fix it here but it can appear elsewhere (?? don't know RR to be sure). We should find at least attempt to find different solution. I can do it.

**Comment by sberyozkin** [comment]
We should probably have this PR closed and continue the discussion in the issue, as any response filters trying to add some header would be affected.

**Comment by FroMage** [comment]
I confirm that I can now use the normal `ExceptionMapper`, thanks a lot, we can close this :)

**Pull request closed without merging**